### PR TITLE
Feature/size and widgets

### PR DESCRIPTION
**Pull request by vanyle**
We now use widgets to display elements inside the code blocks.

This fixes multiple resizing bugs, makes the resizable area easier to change and gives better resizing for the python editor.

Moreover, this hugely improves the code quality by removing huge chunks of unused code.

### CODE REVIEW

**Comment by MathisFederico** [comment]
A few points while testing
- Resizing corner does not exactly follow the mouse
- (Optional) When resizing corner, the code size should be constant

File: opencodeblocks/graphics/blocks/block.py [hunk 1]
```diff
@@ -94,49 +114,50 @@ def boundingRect(self) -> QRectF:
...
+
+        Args:
+            color: title color.
+            font: title font.
+            size: title size.
+            padding: title padding.
+
+        """
+        self.title_widget.setMargin(int(padding))
+        self.title_widget.setStyleSheet("QLabel { color : %s; }" % color)
```
Comment (MathisFederico) [inline opencodeblocks/graphics/blocks/block.py:129]: Use f-string instead
```suggestion
        self.title_widget.setStyleSheet(f"QLabel \{ color : {color}\}")
```

File: opencodeblocks/graphics/pyeditor.py [hunk 2]
```diff
@@ -85,7 +85,7 @@ def update_theme(self):
 
     def views(self) -> List['OCBView']:
         """ Get the views in which the python_editor is present. """
-        return self.graphicsProxyWidget().scene().views()
+        return self.parent().parent().graphicsProxyWidget().scene().views()
```
Comment (MathisFederico) [inline opencodeblocks/graphics/pyeditor.py:88]: Maybe we should have a self.block attribute instead

File: opencodeblocks/graphics/scene/scene.py [hunk 3]
```diff
@@ -38,6 +38,8 @@ def __init__(self, parent=None,
         self.grid_size = grid_size
         self.grid_squares = grid_squares
 
+        self.children = [] # List of widgets
```
Comment (MathisFederico) [inline opencodeblocks/graphics/scene/scene.py:41]: Is that used ?

**Review by MathisFederico — CHANGES_REQUESTED** [review]
A few things to fix, but a very nice refactor of blocks !

File: opencodeblocks/graphics/blocks/blocksizegrip.py [hunk 4]
```diff
@@ -6,22 +6,23 @@
...
 class BlockSizeGrip(QSizeGrip):
-    def __init__(self, container: QGraphicsItem, parent: QWidget = None):
+    def __init__(self, block: QGraphicsItem, parent: QWidget = None):
         """
             Constructor for BlockSizeGrip
 
-            container is the QGraphicsItem holding the QSizeGrip.
-            We should have: parent.graphicsProxyWidget().parent() == container
+            block is the QGraphicsItem holding the QSizeGrip.
+            It's usually an OCBBlock
```
Comment (MathisFederico) [inline opencodeblocks/graphics/blocks/blocksizegrip.py:20]: Why don't you type check with OCBBlock then ?

File: opencodeblocks/graphics/blocks/blocksizegrip.py [hunk 5]
```diff
@@ -36,25 +37,30 @@ def mouseReleaseEvent(self, mouseEvent: QMouseEvent):
 
     def mouseMoveEvent(self, mouseEvent: QMouseEvent):
         """ Performs resizing of the root widget """
+        transformed_pt1 = self.block.mapFromScene(QPoint(0,0))
+        transformed_pt2 = self.block.mapFromScene(QPoint(1,1))
 
-        deltaX = mouseEvent.globalX() - self.mouseX
-        deltaY = mouseEvent.globalY() - self.mouseY
+        pt = transformed_pt2 - transformed_pt1
+        pt /= self.block.scene().views()[0].zoom
```
Comment (MathisFederico) [inline opencodeblocks/graphics/blocks/blocksizegrip.py:44]: Add self.block.scene().views()[0] as a property self.view

File: opencodeblocks/graphics/blocks/codeblock.py [hunk 6]
```diff
@@ -44,6 +50,28 @@ def init_source_editor(self):
...
+        """ Update the code block parts. """
+        super().update_all()
+        if hasattr(self, 'run_button'):
+            self.run_button.setGeometry(
+                int(self.edge_size),
+                int(self.edge_size + self.title_height),
+                int(2.5*self.edge_size),
+                int(2.5*self.edge_size)
+            )
+
```
Comment (MathisFederico) [inline opencodeblocks/graphics/blocks/codeblock.py]: I don't think this is useful anymore, is it ?

File: opencodeblocks/graphics/blocks/codeblock.py [hunk 7]
```diff
@@ -67,7 +95,13 @@ def stdout(self, value: str):
...
             # text output
             self.image = ""
-            self.display.setText(self._stdout)
+
+            editor_widget = self.display
+            # Remove ANSI color codes
+            text = ansi_escape.sub('', value)
+            # Remove backspaces (tf loading bars)
+            text = text.replace('\x08', '')
+            editor_widget.setText(text)
```
Comment (MathisFederico) [inline opencodeblocks/graphics/blocks/codeblock.py]: Naming is wrong here, self.display is not an 'editor_widget'

**Review by MathisFederico — CHANGES_REQUESTED** [review]
I don't see buttons when running main ...

File: opencodeblocks/graphics/blocks/blocksizegrip.py (hunk 4 above)
Comment (vanyle) [reply opencodeblocks/graphics/blocks/blocksizegrip.py:21]: This type of Grip could be used for other types of container. Currently, it's only used for OCBBlocks

**Review by MathisFederico — CHANGES_REQUESTED** [review]
Testing lead to a few things to change:
- Widget states (slider position, output pannel size, ...) are not serialized / deserialized meaning not saved and reloaded, and not kept when copy / pasting.
- Button should be at the left of the title, it takes more space when behind the number line.

**Review by MathisFederico — APPROVED** [review]
Looks good now !

**Pull request merged**